fetchai/ledger/parser/etch_parser.py

```diff
@@ -0,0 +1,259 @@
+from lark import Lark, tree, lexer
+from os import path
+
+# Utility function for checking if a node exists within a (sub)tree
+
+
+def tree_contains(tree_root: tree.Tree, tree_node: tree.Tree):
+    return tree_node in tree_root.iter_subtrees_topdown()
+
+
+class AnnotatedTree:
+    def __init__(self, parsed_tree: tree.Tree):
+        pass
+
+
+class Function:
+    def __init__(self, parsed_tree: tree.Tree):
+        pass
+
+
+class ShardUse:
+    """Class for representing and parsing legal persistent global shard use elements"""
+
+    def __init__(self):
+        self.value = ""
+
+    @staticmethod
+    def from_tree(node: tree.Tree, parameter_dict):
+        assert node.data == 'use_shard_item'
+        if isinstance(node.children[0], lexer.Token):
+            if node.children[0].type == 'NAME':
+                return parameter_dict[node.children[0].value]
+            elif node.children[0].type == 'STRING_LITERAL':
+                return StringLiteral(node.children[0].value)
+        elif isinstance(node.children[0], tree.Tree):
+            # String concat
+            items = []
+            for t in node.children[0].children:
+                if t.type == 'NAME':
+                    items.append(parameter_dict[t.value])
+                elif t.type == 'STRING_LITERAL':
+                    items.append(StringLiteral(t.value))
+            return StringConcat(items)
+
+    def inject_parameters(self, par_dict):
+        pass
+
+
+class Parameter(ShardUse):
+    def __init__(self, name, ptype, value=None):
+        super().__init__()
+        self.name = name
+        self.ptype = ptype
+        self.value = value
+
+    def inject_parameters(self, par_dict):
+        self.value = par_dict[self.name]
+
+
+class StringLiteral(ShardUse):
+    def __init__(self, value=""):
+        super().__init__()
+        self.value = value.strip("\'\"")
+
+
+class StringConcat(ShardUse):
+    def __init__(self, items: list):
+        super().__init__()
+        for i in items:
+            assert isinstance(i, ShardUse)
+        self.items = items
+
+    @property
+    def value(self):
+        return ''.join(i.value for i in self.items)
+
+    @value.setter
+    def value(self, v):
+        pass
+
+    def inject_parameters(self, par_dict):
+        for i in self.items:
+            i.inject_parameters(par_dict)
+
+
+class PersistentGlobal:
+    def __init__(self, name, is_sharded=False, gtype=None):
+        self.name = name
+        self.is_sharded = is_sharded
+        self.gtype = gtype
+
+
+class EtchParser:
+    def __init__(self, etch_code=None):
+        # Load grammar
+        script_dir = path.dirname(path.realpath(__file__))
```

Comment:
It would be worth checking here that this still works after we have packaged this module up into a wheel and then installed on the client system

Reply:
I've replaced this with the pkg_resources module, which should be more robust. I've also added etch.grammar to the list of files that setup.py packages up. This still untested though - what would be the best way to 'install' this for testing?